### PR TITLE
Fixed #33788 -- Added TrigramStrictWordSimilarity() and TrigramStrictWordDistance() on PostgreSQL.

### PR DESCRIPTION
**Pull request by mb1511**
[Ticket #33788](https://code.djangoproject.com/ticket/33788#ticket)

This PR adds support for the PostgreSQL `strict_word_similarity` function, `<<<->` & `%>>` operators. This builds on work done in #14833.

This feature requires the `pg_tgrm` extension to be enabled and requires PostgreSQL 11+. I've run tests on Ubuntu 22 LTS Postgres 12.

### CODE REVIEW

File: django/contrib/postgres/search.py [hunk 1]
```diff
@@ -365,6 +365,16 @@ class TrigramWordDistance(TrigramWordBase):
...
+        super().__init__(string, expression, **extra)
+
 
 class TrigramWordSimilarity(TrigramWordBase):
     function = "WORD_SIMILARITY"
+
+    def __int__(self, string, expression, strict=False, **extra):
+        if strict:
+            self.function = "STRICT_WORD_SIMILARITY"
+        super().__init__(string, expression, **extra)
```
Comment (felixxm) [inline django/contrib/postgres/search.py:380]: The only thing that we change  is a `function`/`arg_joiner` so I would add a separate classes:
```python
class TrigramStrictWordDistance(TrigramWordBase):
    function = ""
    arg_joiner = " <<<-> "


class TrigramStrictWordSimilarity(TrigramWordBase):
    function = "STRICT_WORD_SIMILARITY"
```

File: docs/ref/contrib/postgres/lookups.txt [hunk 2]
```diff
@@ -48,6 +48,20 @@ The ``trigram_word_similar`` lookup can be used on
...
+-------------------------------
+
+.. fieldlookup:: trigram_strict_word_similar
+
+Similar to ``trigram_word_similar``, the ``trigram_strict_word_similar`` lookup
+allows you to perform word similarity lookups using the PostgreSQL extension.
+It differs to ``trigram_word_similar`` by forcing extent boundaries to match
+word boundaries.
+
+Requires PostgreSQL >= 11.0.
```
Comment (felixxm) [inline docs/ref/contrib/postgres/lookups.txt:61]: Django 4.2+ support PostgreSQL 12+, so this note is unnecessary.

File: docs/ref/contrib/postgres/lookups.txt [hunk 3]
```diff
@@ -48,6 +48,20 @@ The ``trigram_word_similar`` lookup can be used on
 
 .. _`pg_trgm extension`: https://www.postgresql.org/docs/current/pgtrgm.html
 
+``trigram_strict_word_similar``
+-------------------------------
```
Comment (felixxm) [inline docs/ref/contrib/postgres/lookups.txt:52]: A new paragraph is unnecessary and it's inconsistent with the current structure of this doc.

File: tests/postgres_tests/__init__.py [hunk 4]
```diff
@@ -21,3 +21,10 @@ class PostgreSQLTestCase(TestCase):
 @modify_settings(INSTALLED_APPS={"append": "django.contrib.postgres"})
 class PostgreSQLWidgetTestCase(WidgetTest, PostgreSQLSimpleTestCase):
     pass
+
+
+requires_postgres_11_up = unittest.skipUnless(
+    connection.vendor == "postgresql"
+    and connection.cursor().connection.server_version >= 110000,
+    "PostgreSQL 11+ specific tests",
+)
```
Comment (felixxm) [inline tests/postgres_tests/__init__.py:30]: Django 4.2+ support PostgreSQL 12+, so this is unnecessary.

File: django/contrib/postgres/apps.py [hunk 5]
```diff
@@ -65,13 +74,8 @@ def ready(self):
...
-        TextField.register_lookup(Unaccent)
-        CharField.register_lookup(SearchLookup)
-        TextField.register_lookup(SearchLookup)
-        CharField.register_lookup(TrigramSimilar)
-        TextField.register_lookup(TrigramSimilar)
-        CharField.register_lookup(TrigramWordSimilar)
-        TextField.register_lookup(TrigramWordSimilar)
+        for lookup in PG_LOOKUPS:
+            CharField.register_lookup(lookup)
+            TextField.register_lookup(lookup)
```
Comment (felixxm) [inline django/contrib/postgres/apps.py:79]: Please revert this unrelated cleanup and add:
```
        CharField._unregister_lookup(TrigramStrictWordSimilar)
        TextField._unregister_lookup(TrigramStrictWordSimilar)
```
```
        CharField.register_lookup(TrigramStrictWordSimilar)
        TextField.register_lookup(TrigramStrictWordSimilar)
```
directly.

**Review by felixxm — COMMENTED** [review]
@mb1511 Thanks for this patch :+1: I left initial comments.

**Comment by mb1511** [comment]
@felixxm Thanks for the quick feedback! :+1: I've addressed all your initial comments.
Also, it's super useful to know 4.2 now only officially supports PG12+ :raised_hands:

**Comment by felixxm** [comment]
@mb1511 Thanks :+1: Welcome aboard :boat: 

I pushed small edits to docs.
